drivers/adc/adc_stm32.c

```diff
@@ -402,6 +402,41 @@ static void adc_stm32_oversampling(ADC_TypeDef *adc, uint8_t ratio, uint32_t shi
 }
 #endif /* CONFIG_SOC_SERIES_STM32xxx */
 
+/*
+ * Enable ADC peripheral, and wait until ready if required by SOC.
+ */
+static int adc_stm32_enable(ADC_TypeDef *adc)
+{
+	LL_ADC_Enable(adc);
+
+#if defined(CONFIG_SOC_SERIES_STM32L4X) || \
+	defined(CONFIG_SOC_SERIES_STM32L5X) || \
+	defined(CONFIG_SOC_SERIES_STM32WBX) || \
+	defined(CONFIG_SOC_SERIES_STM32G0X) || \
+	defined(CONFIG_SOC_SERIES_STM32G4X) || \
+	defined(CONFIG_SOC_SERIES_STM32H7X) || \
+	defined(CONFIG_SOC_SERIES_STM32WLX)
+	/*
+	 * Enabling ADC modules in L4, WB, G0 and G4 series may fail if they are
+	 * still not stabilized, this will wait for a short time to ensure ADC
+	 * modules are properly enabled.
+	 */
+	uint32_t count_timeout = 0;
+
+	while (LL_ADC_IsActiveFlag_ADRDY(adc) == 0) {
+		if (LL_ADC_IsEnabled(adc) == 0UL) {
+			LL_ADC_Enable(adc);
+			count_timeout++;
+			if (count_timeout == 10) {
+				return -ETIMEDOUT;
+			}
+		}
+	}
+#endif
+
```

Comment:
should the ADC Ready bit ADRDY bit be cleared afterwards ?

Reply:
I just looked in RM0453. Under 18.3.4 ADC on-off control (ADEN, ADDIS, ADRDY) it says to clear ADRDY as _before_ enabling and (optionally) after _disabling_.

It's cleared by software so obviously it has to be cleared sometime or the check is pointless. Currently this is done in `adc_stm32_init` before enabling the ADC, but it's not apparent from the code if this is the intent.

Reply:
I had a look in a few other reference manuals and most of them agree that ADRDY should be cleared before enabling, RM0456 says to clear it afterwards as well, RM0316 doesn't mention clearing it although that seems like an oversight. Every manual that I checked says to wait for ADRDY after enabling, but this is not done for all relevant series at the moment.

I'm not sure how to proceed. My aim was to fix an issue that I was having with STM32WL. I believe this PR can serve as a first step toward improving the handling of ADRDY but I unfortunately don't have resources to fix it for every case.

Reply:
Are the tests available in zephyr sufficient to test this use case ?
If this is the case, I propose we test the current change on our board farm.

Reply:
@abxy, @FRASTM can you comment my proposal?

Reply:
@erwango What do you mean by "this use case"? Are you referring to the discussion above, or to the failure I have encountered? The sample `samples/drivers/adc` is sufficient to test the failure I'm having. Every sample read is 0.

```
** Booting Zephyr OS build v2.7.99-2866-g4963959afc12  ***
ADC reading: 0
ADC reading: 0
ADC reading: 0
```

Inspecting the `handler.log` for `tests/drivers/adc/adc_api` running on target it consistently reads a 0 for `test_adc_sample_one_channel` before my change, but this is not considered a failure. The other tests in the same suite don't read all 0's, but they all seem to be doing repeated samples in one way or another.

Command to run test on target:
```
twister --device-testing --device-serial /dev/ttyACM0 --device-serial-baud 115200 -p nucleo_wl55jc  -T tests/drivers/adc/adc_api/
```

Reply:
@abxy I meant the ADRDY handling, if we attempt to change it on all series (assuming we'll be able to test on all series)

Reply:
@FRASTM  I had a check in a bunch more reference manuals and I think the best approach is to clear ADRDY before enabling. I added a separate commit for this change.

@erwango What type of test can you run on your board farm? Every series affected by the ADRDY handling (L4, L5, WB, G0, G4, H7, WL) is covered by a Nucleo board configuration enabled in `tests/drivers/adc/adc_api`. Unfortunately the test itself passes a zero valued sample which means it won't catch all errors related to this.

The `dac_loopback` test would be better suited I think, but it doesn't seem to have coverage for all series. I'm trying to get it to run on my board to test if it can trigger my original problem.

Reply:
> What type of test can you run on your board farm?

All tests in tree tests. We also have (or will add if missing) fixtures for things like loopback tests.

> The dac_loopback test would be better suited I think, but it doesn't seem to have coverage for all series.

We can extend coverage as long as dac is available on these series.

Reply:
For some reason `tests/drivers/dac_loopback` doesn't trigger this issue for me. I don't understand exactly why. I can modify the test it to get the issue to trigger, but small changes can make it disappear again. The adc sample on the other hand shows the issue reliably even, when I add sleeps of varying length between calls to the adc api.